File: identity-android/src/main/java/com/android/identity/android/mdoc/deviceretrieval/IsoDepWrapperImpl.java
/*
 * Copyright 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.identity.android.mdoc.deviceretrieval;

import android.nfc.Tag;
import android.nfc.tech.IsoDep;
import java.io.IOException;

public class IsoDepWrapperImpl implements IsoDepWrapper {
  private IsoDep mIsoDep;

  public IsoDepWrapperImpl(Tag tag) {
    mIsoDep = IsoDep.get(tag);
  }

  @Override
  public Tag getTag() {
    return mIsoDep.getTag();
  }

  @Override
  public boolean isConnected() {
    return mIsoDep.isConnected();
  }

  @Override
  public void connect() throws IOException {
    mIsoDep.connect();
  }

  @Override
  public void close() throws IOException {
    mIsoDep.close();
  }

  @Override
  public void getIsoDep(Tag tag) {
    mIsoDep = IsoDep.get(tag);
  }

  @Override
  public boolean isTagSupported() {
    return mIsoDep != null;
  }

  @Override
  public void setTimeout(int timeout) {
    mIsoDep.setTimeout(timeout);
  }

  @Override
  public int getTimeout() {
    return mIsoDep.getTimeout();
  }

  @Override
  public byte[] getHistoricalBytes() {
    return mIsoDep.getHistoricalBytes();
  }

  @Override
  public byte[] getHiLayerResponse() {
    return mIsoDep.getHiLayerResponse();
  }

  @Override
  public int getMaxTransceiveLength() {
    // This value is set based on the Pixel's eSE APDU Buffer size
    return 261;
  }

  @Override
  public byte[] transceive(byte[] data) throws IOException {
    return mIsoDep.transceive(data);
  }

  @Override
  public boolean isExtendedLengthApduSupported() {
    return mIsoDep.isExtendedLengthApduSupported();
  }
}

File: multipaz-android-legacy/src/main/java/com/android/identity/android/mdoc/deviceretrieval/IsoDepWrapperImpl.kt
package com.android.identity.android.mdoc.deviceretrieval

import android.nfc.Tag
import android.nfc.tech.IsoDep
import java.io.IOException


class IsoDepWrapperImpl(tag: Tag?) : IsoDepWrapper {
    private var mIsoDep: IsoDep? = null

    init {
        mIsoDep = IsoDep.get(tag)
    }

    override fun getTag(): Tag {
        return mIsoDep!!.tag
    }

    override val isConnected: Boolean
        get() {
            return mIsoDep!!.isConnected
        }

    @Throws(IOException::class)
    override fun connect() {
        mIsoDep!!.connect()
    }

    @Throws(IOException::class)
    override fun close() {
        mIsoDep!!.close()
    }

    override fun getIsoDep(tag: Tag?) {
        mIsoDep = IsoDep.get(tag)
    }

    override val isTagSupported: Boolean
        get() {
            return mIsoDep != null
        }

    override fun setTimeout(timeout: Int) {
        mIsoDep!!.timeout = timeout
    }

    override fun getTimeout(): Int {
        return mIsoDep!!.timeout
    }

    override fun getHistoricalBytes(): ByteArray {
        return mIsoDep!!.historicalBytes
    }

    override fun getHiLayerResponse(): ByteArray {
        return mIsoDep!!.hiLayerResponse
    }

    override val maxTransceiveLength: Int
        get() {
            // This value is set based on the Pixel's eSE APDU Buffer size
            return 261
        }

    @Throws(IOException::class)
    override fun transceive(data: ByteArray?): ByteArray {
        return mIsoDep!!.transceive(data)
    }

    override val isExtendedLengthApduSupported: Boolean
        get() {
            return mIsoDep!!.isExtendedLengthApduSupported
        }
}